pgxn/neon/neon.c

```diff
@@ -267,6 +268,179 @@ LogicalSlotsMonitorMain(Datum main_arg)
 	}
 }
 
+/*
+ * This is taken from procarry.c
+ * TODO: should we better move it to some header file?
+ */
+#define PROCARRAY_MAXPROCS	(MaxBackends + max_prepared_xacts)
+#define TOTAL_MAX_CACHED_SUBXIDS \
+	((PGPROC_MAX_CACHED_SUBXIDS + 1) * PROCARRAY_MAXPROCS)
+
+/*
+ * We do not want to wait running-xacts record from primary which is
+ * generated each 15 seconds (if there is some activity).  So we try
+ * to restore information about running transactions from CLOG.
+ */
+/*
+ * NEON: we do not want to wait running-xacts record from primary which is generated each 15 seconds (if there is some activity).
+ * So we try to restore information about running transactions from CLOG.
+ *
+ * This functions allocates and fill array at *xids and store number of restored xacts in nxids.
+ * In case of overflow, false is returned, *nxids == 0 and *xids == NULL.
+ */
+static bool
+RestoreRunningXactsFromClog(CheckPoint *checkpoint, TransactionId **xids, int *nxids)
+{
+	int			xcnt;
+	int			max_xcnt;
+	TransactionId* restored_xids;
+	TransactionId xid;
+	XLogRecPtr	xidlsn;
+	XidStatus	xidstatus;
+	bool		overflow = false;
+	int			n_prepared_xacts;
+
+	TransactionId from = TransactionIdPrecedes(checkpoint->oldestXid, checkpoint->oldestActiveXid)
+		? checkpoint->oldestActiveXid : checkpoint->oldestXid;
```

Comment:
This desperately needs a comment. Why can oldestXid ever be ahead of oldestActiveXid? That would imply that we already truncate the CLOG for a transaction that's still in progress.

If I understand correctly, it's because oldestActiveXid is updated in the pageserver whenever it ingests a checkpoint record, and oldestXid is updated whenever it sees a checkpoint record or a CLOG_TRUNCATE record. So for example, it's possible that XID 123 commits, then autovacuum runs and truncates the CLOG. The oldestActiveXid tracked in the pageserver is still 123, but ingesting the CLOG_TRUNCATE record advances oldestXid past that.

Would it be better to also advance oldestActiveXid in `ingest_clog_truncate_record`?

There are some comments in `ingest_clog_truncate_record` on how checkpoint.oldestXid lags behind on Postgres, and how it's just an optimization that we advance it. That seems outdated now.